client_test/directfund_simple_tcp_test.go

```diff
@@ -0,0 +1,46 @@
+// Package client_test contains helpers and integration tests for go-nitro clients
+package client_test // import "github.com/statechannels/go-nitro/client_test"
+
+import (
+	"io"
+	"testing"
+	"time"
+
+	"github.com/statechannels/go-nitro/client"
+	"github.com/statechannels/go-nitro/client/engine/chainservice"
+	simpletcp "github.com/statechannels/go-nitro/client/engine/messageservice/simple-tcp"
+	"github.com/statechannels/go-nitro/client/engine/store"
+	"github.com/statechannels/go-nitro/crypto"
+	"github.com/statechannels/go-nitro/types"
+)
+
+// setupClientWithSimpleTCP is a helper function that contructs a client and returns the new client and its store.
+func setupClientWithSimpleTCP(pk []byte, chain chainservice.MockChain, peers map[types.Address]string, logDestination io.Writer, meanMessageDelay time.Duration) (client.Client, *simpletcp.SimpleTCPMessageService) {
```

Comment:
How can we organise our integration tests, such that we aren't repeating ourselves? We probably don't want to write a new test for every (protocol, message service, store, chain service) combination. At least not manually.

Perhaps if we were careful enough with the test setup /teardown interfaces, we could actually programmatically try out all of the combinations? (I realise the setup and teardown are currently a little different between the two message services).

Another option would be just to drop this TCP service in over the existing `chan` based service in all of our integration tests. A did some quick experiments and it seems to be just as fast. I wonder if there is much value keeping the `chan` based service around, since the TCP one is probably more realistic. What are your thoughts?

Reply:
I think switching to the TCP message service would make sense. I'd be in favour of just deprecating the `chan` based service so we only have to maintain the one message service if we could.  We'd need to add message delay logic for the message delay test and probably want to create a vectorclock version as well?

I think by default tests in a package are [not run in parallel](https://stackoverflow.com/questions/24375966/does-go-test-run-unit-tests-concurrently) so I think we'd avoid messages services fighting over a TCP port, but it is something to keep in mind.

In general I think we could also refactor some of our integration tests into one file. Currently the `virtualfund_test`, `virtualfund_with_message_delay` and `virtualfund_multiparty_test` all have their own test files but they're essentially similar tests with some different parameters passed in.